src/d3a/models/area/markets.py

```diff
@@ -15,84 +15,175 @@
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-from pendulum import DateTime # noqa
-from typing import Dict  # noqa
-
-from d3a.models.market import GridFee
-from d3a.models.market.two_sided import TwoSidedMarket
-from d3a.models.market.one_sided import OneSidedMarket
-from d3a.models.market.balancing import BalancingMarket
-from d3a.models.market import Market # noqa
-from d3a_interface.constants_limits import ConstSettings
+from abc import ABC
 from collections import OrderedDict
+from logging import getLogger
+from typing import Dict, TYPE_CHECKING
+
+from d3a import constants
 from d3a.d3a_core.util import is_timeslot_in_simulation_duration
-import d3a.constants
+from d3a.models.market import GridFee, Market
+from d3a.models.market.balancing import BalancingMarket
+from d3a.models.market.one_sided import OneSidedMarket
+from d3a.models.market.two_sided import TwoSidedMarket
+from d3a_interface.constants_limits import ConstSettings, GlobalConfig
+from pendulum import DateTime
+
+if TYPE_CHECKING:
+    from d3a.models.area import Area
+
+log = getLogger(__name__)
+
+
+class RotatorInterface(ABC):
+    """Market rotator interface"""
+
+    def rotate(self, current_time_slot):
+        pass
+
+
+class MarketRotatorBase(RotatorInterface):
+    """Deal with market rotation."""
+
+    def __init__(self, markets: Dict, past_markets: Dict):
+        self.markets = markets
+        self.past_markets = past_markets
+
+    def rotate(self, current_time_slot: DateTime) -> None:
+        """Move markets to past and delete old past markets."""
+        self._move_markets_to_past(self.markets, self.past_markets, current_time_slot)
+        self._delete_past_markets(self.past_markets, current_time_slot)
+
+    @staticmethod
+    def _is_it_time_to_delete_past_market(current_time_slot: DateTime,
+                                          time_slot: DateTime) -> bool:
+        """Check if the past market for time_slot is ready to be deleted."""
+
+        if constants.RETAIN_PAST_MARKET_STRATEGIES_STATE:
+            return False
+        else:
+            if ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKETS:
+                return (time_slot < current_time_slot.subtract(
+                    hours=ConstSettings.GeneralSettings.MAX_AGE_SETTLEMENT_MARKET_HOURS))
```

Comment:
Since we have a specialized rotator for each market type, these `if` statements can be implemented as methods of those classes. E.g.:
```python
class MarketRotatorBase:
    # ...
    def is_it_time_to_delete_past_market(current_time_slot: DateTime, time_slot: DateTime) -> bool:
        if constants.RETAIN_PAST_MARKET_STRATEGIES_STATE:
            return False

        return time_slot < current_time_slot.subtract(
            minutes=GlobalConfig.slot_length.total_minutes())

class SettlementMarketRotator(MarketRotatorBase):
    # ...
    def is_it_time_to_delete_past_market(current_time_slot: DateTime, time_slot: DateTime) -> bool:
        if constants.RETAIN_PAST_MARKET_STRATEGIES_STATE:
            return False

        if ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKETS:
            return (time_slot < current_time_slot.subtract(
                hours=ConstSettings.GeneralSettings.MAX_AGE_SETTLEMENT_MARKET_HOURS))
```
There's some duplication in my example that could probably be further avoided.

Reply:
@fievelk we need to keep more past markets in the memory in case the settlement markets are enabled. This is one reason I decided to go for the code dublication in my first proposal.

Reply:
I think my suggestion above shouldn't change what you described, right? `_delete_past_markets()` will call `self. _is_it_time_to_delete_past_market()`. At that point, if you're inside a `SettlementMarketRotator` instance you will keep more past markets in memory, otherwise if you're in a `MarketRotatorBase` instance you'll keep the "normal" amount (based on the method implementation).

Reply:
I meant that we also need to keep more SPOT past markets in the memory if `ConstSettings.GeneralSettings.ENABLE_SETTLEMENT_MARKET`. So, the `MarketRotatorBase. is_it_time_to_delete_past_market` should return 
```python
            return (time_slot < current_time_slot.subtract(
                hours=ConstSettings.GeneralSettings.MAX_AGE_SETTLEMENT_MARKET_HOURS))
```
in this case.

Reply:
Interesting, I hadn't understood that. Thanks for the explanation. Maybe a comment would help, if you have time to add it.

Reply:
Added it https://github.com/gridsingularity/d3a/pull/1213/commits/983054e368a590c99a8e6e6bd37f5213200b3ed3
Will press push after the tests have passed.

Reply:
Great, thanks!